### PR TITLE
Remove forced whodata mode reporting in Windows

### PR DESCRIPTION
**Pull request by crolopez**
This PR solves the issue described in #3599 and #3582 for whodata in Windows. 

By design, and whenever possible, all FIM events reported by Syscheck will have whodata information. This is achieved by checking for each processed patch if it has whodata information.

The problem appears when a path, whose closest parent does not have whodata, has a more distant parent with whodata enabled.

This path can be reported with whodata information because it has inherited the audit policies, but the results with certain configurations may not working as expected.

As explained in the issue, the problem is the deep search, so this PR removes it so that each path is processed with the attributes of its nearest directory, regardless of whether it has inherited audit policies and has extra information.

## Possible invalid behaviours

The issue can be replied with only two monitored directories, being one of them parent to the other and being monitored in whodata mode.

In the following example, the events from the subfolders will be reported in realtime with audit information. This is not the expected behaviour because the subfolder has to be scanned every `frequency` period.

``` XML
<directories check_all="yes" whodata="yes">C:\test_folder</directories>
<directories check_all="yes">C:\test_folder\subfolder</directories>
```

The following example will work as expected because the realtime mode is faster than the whodata mode, so the files reported in the subfolder will not have audit information. In spite of that, it is an invisible race condition.

``` XML
<directories check_all="yes" whodata="yes">C:\test_folder</directories>
<directories check_all="yes" realtime="yes">C:\test_folder\subfolder</directories>
```

If both directories are monitored with whodata, may seem to report correctly, but they do not. If we apply a specific configuration for the subfolder, we will never see it. In this example, subfolder events should include `tag1` in the alert, but include `tag0` by the less restrictive parent with whodata enabled.

``` XML
<directories check_all="yes" whodata="yes" tags="tag0">C:\test_folder</directories>
<directories check_all="yes" whodata="yes" tags="tag1">C:\test_folder\subfolder</directories>
```

## Tags only work in added events

If the monitored folder (in whodata mode) is not the first configured directory, its reports will only display the correct tags in added events, regardless of the directories scope. For example:

``` XML
<directories check_all="yes" tags="tag0">C:\another\test\folder</directories>
<directories check_all="yes" whodata="yes" tags="tag1">C:\test_folder\subfolder</directories>
```

If a modification event is generated on `C:\test_folder\subfolder\file.txt` with the previous configuration, it will have the attributes of whodata but will include the tag `tag0`, which belongs to a directory without whodata and without parent relation, because the position of the directory is not being properly controlled in these operations. The PR also solves this issue.

A good way to check that this PR fixes the issue is to move a file from the child directory to the parent directory. An alert should appear for each directory with their respective tags. For example:

``` XML
<directories check_all="yes" whodata="yes" tags="tag0">C:\test_folder</directories>
<directories check_all="yes" whodata="yes" tags="tag1">C:\test_folder\subfolder</directories>
```

> ** Alert 1562062628.342679: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
> 2019 Jul 02 10:17:08 (agwin2016) any->syscheck
> Rule: 554 (level 5) -> 'File added to the system.'
> File 'c:\test_folder\file.txt' was added.
> 
> Attributes:
>  - Size: 22
>  - Date: Tue Jul  2 10:00:26 2019
>  - User: Administrators (S-1-5-32-544)
>  - MD5: 76cdb2bad9582d23c1f6f4d868218d6c
>  - SHA1: b04f3ee8f5e43fa3b162981b50bb72fe1acabb33
>  - SHA256: 8739c76e681f900923b900c9df0ef75cf421d39cabb54650c4b9ad19b6a76d85
>  - File attributes: ARCHIVE
>  - Permissions: 
>    SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
>    Administrators  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
>    Administrator  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
> 
> Tags:
>  - tag0

> 
> ** Alert 1562062628.344165: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
> 2019 Jul 02 10:17:08 (agwin2016) any->syscheck
> Rule: 553 (level 7) -> 'File deleted.'
> File 'c:\test_folder\subfolder\file.txt' was deleted.
> (Audit) User: 'Administrator (S-1-5-21-3292556202-24657078-706277677-500)'
> (Audit) Process id: '4084'
> (Audit) Process name: 'C:\Windows\explorer.exe'
> 
> 
> Tags:
>  - tag1

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [ ] MAC OS X
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- Memory tests
  - [x] Valgrind report for affected components
  - [x] CPU impact
  - [x] RAM usage impact
- [x] Retrocompatibility with older Wazuh versions
- [x] Working on cluster environments

### CODE REVIEW

**Comment by bah07** [comment]
Hi team,

There is another task regarding this problem, in case of adding this configuration, 
```
  <!-- File integrity monitoring -->
  <syscheck>

    <frequency>400</frequency>
    <directories check_all="yes" recursion_level="0" check_mtime="no" report_changes="no" whodata="yes" tags="firsttag">C:\test</directories>
    <directories check_all="yes" recursion_level="0" check_mtime="no" report_changes="yes" whodata="yes" tags="secondtag">C:\test\f1\f2>directories>

  </syscheck>

```
no ADD alert is received when a new file is generated. The expected are 3 alerts since Windows adds the file first with a default name "new document.txt".  

First:
```
** Alert 1562074291.2027550: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
2019 Jul 02 15:31:31 (WIN-3SGVSBG9BN4) 10.0.0.5->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'c:\test\f1\f2\nuevo documento de texto.txt' was added.
(Audit) User: 'Administrador (S-1-5-21-2668198274-3253688744-2454179532-500)'
(Audit) Process id: '3040'
(Audit) Process name: 'C:\Windows\explorer.exe'

Attributes:
 - Size: 0
 - User: Administradores (S-1-5-32-544)
 - MD5: d41d8cd98f00b204e9800998ecf8427e
 - SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
 - SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
 - File attributes: ARCHIVE
 - Permissions: 
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES

Tags:
 - secondtag

```
Second:
```
** Alert 1562074292.2028832: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
2019 Jul 02 15:31:32 (WIN-3SGVSBG9BN4) 10.0.0.5->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'c:\test\f1\f2\ffff.txt' was added.

Attributes:
 - Size: 0
 - User: Administradores (S-1-5-32-544)
 - MD5: d41d8cd98f00b204e9800998ecf8427e
 - SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
 - SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
 - File attributes: ARCHIVE
 - Permissions: 
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES

Tags:
 - secondtag

```
Third:
```
** Alert 1562074293.2029941: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
2019 Jul 02 15:31:33 (WIN-3SGVSBG9BN4) 10.0.0.5->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'c:\test\f1\f2\nuevo documento de texto.txt' was deleted.
(Audit) User: 'Administrador (S-1-5-21-2668198274-3253688744-2454179532-500)'
(Audit) Process id: '3040'
(Audit) Process name: 'C:\Windows\explorer.exe'

Tags:
 - secondtag
```

In this use case, the first ADD and DELETE alerts appear with respect to the name given by Windows in the creation but the ADD alert does not appear when renaming the file.

I think that the problem are related with the `recursion_level` option.

**Comment by crolopez** [comment]
Fixed in 939cb204b81ac9fdac7c72dee183064373c7fcce.

The reason for the bug was that the difference between the parent directory and the new file was not being increased. The solution is to add `+ 1` to this difference:

https://github.com/wazuh/wazuh/blob/939cb204b81ac9fdac7c72dee183064373c7fcce/src/syscheckd/win_whodata.c#L825

**Comment by bah07** [comment]
Hi team,

We need to fix this issue #3425 and we decide to solve in this PR.

Commit: 4aef1e1299988e1c56f1ff078cb4edb0c95f5536

**Comment by bah07** [comment]
Valgrind test:
```
==30692== LEAK SUMMARY:
==30692==    definitely lost: 0 bytes in 0 blocks
==30692==    indirectly lost: 0 bytes in 0 blocks
==30692==      possibly lost: 816 bytes in 3 blocks
==30692==    still reachable: 8,861,766 bytes in 118,451 blocks
==30692==         suppressed: 0 bytes in 0 blocks
==30692== 

```

**Comment by DaveVG1** [comment]
While testing this branch I've found the following: there's a lack of delete alert after copying a file into a folder and deleting that folder.
This is my configuration:
```
    <directories check_all="yes" whodata="yes" tags="testfim-wd">D:\Libraries\Escritorio\testfim</directories>
    <directories check_all="yes" realtime="yes" tags="rt1">D:\Libraries\Escritorio\testfim\folder1</directories>
    <directories check_all="yes" whodata="yes" tags="wd2">D:\Libraries\Escritorio\testfim\folder1\folder2</directories>
    <directories check_all="yes" realtime="yes" tags="rt2">D:\Libraries\Escritorio\testfim\folder1\folder2\folder3</directories>
    <directories check_all="yes" whodata="yes" realtime="yes" tags="rt-wd">D:\Libraries\Escritorio\testfim\folder1\folder2\folder3\folder4</directories>
    <directories check_all="yes" tags="nothing">D:\Libraries\Escritorio\testfim\folder1\folder2\folder3\folder4\folder5</directories>
```
These are the steps to reproduce this:
1) In `folder1` create a file. The following alert should appear:
```
** Alert 1562488208.166016: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Jul 07 10:30:08 (Win) 10.0.0.1->syscheck
Rule: 554 (level 5) -> 'File added to the system.'
File 'd:\libraries\escritorio\testfim\folder1\nuevo microsoft excel worksheet.xlsx' was added.

Attributes:
 - Size: 6187
 - Date: Tue Jul 16 17:26:07 2019
 - User: DaveVG (S-1-5-21-4218398017-4192086937-1907420761-1001)
 - MD5: 4afbb1718e49e7a2988c418d0be384ed
 - SHA1: 81ae5f60bba95e8b0b5475ea971817062bb139ea
 - SHA256: d2e4e3aa98008d92348ab088ec3f6c12a1f3863cd05ac41343b59ac1f8d0c141
 - File attributes: ARCHIVE
 - Permissions:
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Authenticated Users  (ALLOWED) - DELETE, READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Authenticated Users  (ALLOWED) - DELETE, READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES

Tags:
 - rt1
```
2) Create a folder name `ha-rt` (for example) inside `folder1`.
3) Drag the previously created file to this new folder. You should get two alerts like these:
```
** Alert 1562488319.168248: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Jul 07 10:31:59 (Win) 10.0.0.1->syscheck
Rule: 553 (level 7) -> 'File deleted.'
File 'd:\libraries\escritorio\testfim\folder1\nuevo microsoft excel worksheet.xlsx' was deleted.


Tags:
 - rt1

** Alert 1562488319.168583: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,hipaa_164.312.c.1,hipaa_164.312.c.2,nist_800_53_SI.7,
2019 Jul 07 10:31:59 (Win) 10.0.0.1->syscheck
Rule: 550 (level 7) -> 'Integrity checksum changed.'
File 'd:\libraries\escritorio\testfim\folder1\ha-rt\nuevo microsoft excel worksheet.xlsx' checksum changed.
Old modification time was: 'Tue Jul 16 17:20:24 2019', now it is 'Tue Jul 16 17:26:07 2019'

Attributes:
 - Size: 6187
 - Date: Tue Jul 16 17:26:07 2019
 - User: DaveVG (S-1-5-21-4218398017-4192086937-1907420761-1001)
 - MD5: 4afbb1718e49e7a2988c418d0be384ed
 - SHA1: 81ae5f60bba95e8b0b5475ea971817062bb139ea
 - SHA256: d2e4e3aa98008d92348ab088ec3f6c12a1f3863cd05ac41343b59ac1f8d0c141
 - File attributes: ARCHIVE
 - Permissions:
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Authenticated Users  (ALLOWED) - DELETE, READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES
   Administradores  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Authenticated Users  (ALLOWED) - DELETE, READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
   Usuarios  (ALLOWED) - READ_CONTROL, SYNCHRONIZE, FILE_READ_DATA, FILE_READ_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES

Tags:
 - rt1
```
4) Delete the previously created folder (named `ha-rt` in the example). No alerts are displayed after doing this.

**Comment by crolopez** [comment]
A configuration as follow produce false positives due to an inconsistency between the realtime watchers and whodata.

``` XML
<directories check_all="yes" tags="FREQ">C:\Users\Administrator\Desktop\test</directories>
<directories check_all="yes" tags="RT0" realtime="yes">C:\Users\Administrator\Desktop\test\realtime0</directories>
<directories check_all="yes" tags="WD0" whodata="yes">C:\Users\Administrator\Desktop\test\realtime0\whodata0</directories>
<directories check_all="yes" tags="RT1" realtime="yes">C:\Users\Administrator\Desktop\test\realtime0\whodata0\realtime1</directories>
<directories check_all="yes" tags="WD1" whodata="yes">C:\Users\Administrator\Desktop\test\realtime0\whodata0\realtime1\whodata1</directories>
<directories check_all="yes" tags="RT2" realtime="yes">C:\Users\Administrator\Desktop\test\realtime0\whodata0\realtime1\whodata1\realtime2</directories>
<directories check_all="yes" tags="WD2" whodata="yes">C:\Users\Administrator\Desktop\test\realtime0\whodata0\realtime1\whodata1\realtime2\whodata2</directories>
```

In this case, a realtime event is reported because it is faster than the whodata mode. It is not correct because the closer parent to the file is in whodata mode. However, the realtime reporting process knows the real parent of the file and generates an alert with its tag.

> ** Alert 1563281534.33008: - ossec,syscheck,pci_dss_11.5,gpg13_4.11,gdpr_II_5.1.f,
> 2019 Jul 16 12:52:14 (agwin2016) any->syscheck
> Rule: 554 (level 5) -> 'File added to the system.'
> File 'c:\users\administrator\desktop\test\realtime0\whodata0\realtime1\whodata1\realtime2\whodata2\subfolder\new text document.txt' was added.
> 
> Attributes:
>  - Size: 0
>  - Date: Tue Jul 16 12:52:14 2019
>  - User: Administrators (S-1-5-32-544)
>  - MD5: d41d8cd98f00b204e9800998ecf8427e
>  - SHA1: da39a3ee5e6b4b0d3255bfef95601890afd80709
>  - SHA256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
>  - File attributes: ARCHIVE
>  - Permissions: 
>    SYSTEM  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
>    Administrators  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
>    Administrator  (ALLOWED) - DELETE, READ_CONTROL, WRITE_DAC, WRITE_OWNER, SYNCHRONIZE, FILE_READ_DATA, FILE_WRITE_DATA, FILE_APPEND_DATA, FILE_READ_EA, FILE_WRITE_EA, FILE_EXECUTE, FILE_READ_ATTRIBUTES, FILE_WRITE_ATTRIBUTES
> 
> Tags:
>  - WD2

In other words, the realtime engine does not check that event generated under its watcher is not for it. On the other hand, whodata for Windows does not have this problem because it evaluates that the closer parent of the processing file may not be for this mode.

https://github.com/wazuh/wazuh/blob/68aaf6d31557106bc8b3cb58b81388644cb27ff5/src/syscheckd/win_whodata.c#L635-L640

In be0663b23dc54484df9729fbfe09197a8a28973c, we have improved this filter at `find_dir_pos` level, which avoids that whodata or realtime process events that are not for them. This also allows us to remove the whodata for Windows own control previously mentioned, and the ignored hash table, reducing code complexity

**Comment by crolopez** [comment]
Under certain deletion conditions, whodata for Windows generated the following warning in Analysisd.

> Agent 'XXX' Couldn't decode fim sum '' from file 'c:\file.txt'.

In spite of this, the corresponding alert was reported. This has been fixed in 94634d90fdee110ba5dbdea3c6b17998c633c4ec.

**Comment by crolopez** [comment]
The FIM events reported through a `read_dir()` call was being generated without proper checks. If in the recursive scan of a realtime folder appears a subfolder monitored without real-time modes, it is notified as with invalid attributes. 

It has been fixed in 10b62a636c72a22d3285f858c545c46e927971b8 and fdc8f095d8a898a85f9171997ea5a3e19a5234db.

**Comment by crolopez** [comment]
## Invalid behaviour of the recursion level

If we set a recursion level, this value will not be taken into account in some cases. To illustrate this, I will use the following configuration:

``` XML
<directories check_all="yes" tags="WD2" whodata="yes" recursion_level="0">C:\folder</directories>
```

If we move a file from `C:\folder` to `C:\folder\subfolder`, we should not see any alert because of the `recursion_level`, but it is happening.

If we combine the previous case with a realtime parent folder that should cover the subfolder, but it still works badly since this event should be reported with realtime and not by whodata.

``` XML
<directories check_all="yes" tags="RT" realtime="yes" recursion_level="2">C:\folder1</directories>
<directories check_all="yes" tags="WD" whodata="yes" recursion_level="0">C:\folder1\folder2</directories>
```

Since ecb5d71b20dfa8928a944b5b7fc56d4b9cbcc2ea, if we move a file from `C:\folder1\folder2` to `C:\folder1\folder2\subfodler`, we will have a remove alert with audit and the `WD` tag and a creation alert from realtime and the `RT` tag.
